lib/usagereporter/teleport/aggregating/service.go

```diff
@@ -160,3 +245,70 @@ func (r reportService) createUserActivityReportsLock(ctx context.Context, ttl ti
 
 	return nil
 }
+
+func (r reportService) upsertResourcePresenceReport(ctx context.Context, report *prehogv1.ResourcePresenceReport, ttl time.Duration) error {
+	wire, err := proto.Marshal(report)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	reportUUID, err := uuid.FromBytes(report.GetReportUuid())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	startTime := report.GetStartTime().AsTime()
+	if startTime.IsZero() {
+		return trace.BadParameter("missing start_time")
+	}
+
+	if _, err := r.b.Put(ctx, backend.Item{
+		Key:     resourcePresenceReportKey(reportUUID, startTime),
+		Value:   wire,
+		Expires: startTime.Add(ttl),
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (r reportService) deleteResourcePresenceReport(ctx context.Context, report *prehogv1.ResourcePresenceReport) error {
+	reportUUID, err := uuid.FromBytes(report.GetReportUuid())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	startTime := report.GetStartTime().AsTime()
+	if startTime.IsZero() {
+		return trace.BadParameter("missing start_time")
+	}
+
+	if err := r.b.Delete(ctx, resourcePresenceReportKey(reportUUID, startTime)); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// listResourcePresenceReports returns the first `count` resource counts reports
+// according to the key order; as we store them with time and uuid in the key,
+// this results in returning earlier reports first.
+func (r reportService) listResourcePresenceReports(ctx context.Context, count int) ([]*prehogv1.ResourcePresenceReport, error) {
+	rangeStart := backend.ExactKey(ResourcePresenceReportsPrefix)
+	result, err := r.b.GetRange(ctx, rangeStart, backend.RangeEnd(rangeStart), count)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reports := make([]*prehogv1.ResourcePresenceReport, 0, len(result.Items))
+	for _, item := range result.Items {
+		report := &prehogv1.ResourcePresenceReport{}
+		if err := proto.Unmarshal(item.Value, report); err != nil {
+			return nil, trace.Wrap(err)
```

Comment:
Potentially, but if we make that change we should do it for user activity reports too, so for now I think it's best to be consistent with current behavior.

Reply:
Failing usage data submission will result in a cluster alert, which will hopefully prompt the customer into calling us. That'd still work if we skipped over invalid data, but we would need to tweak the logic around creating and deleting the cluster alerts (to still create one), since we don't want to keep ignoring some logic bug.